### PR TITLE
ARROW-5182: [Rust] Arrow IPC file writer

### PR DESCRIPTION
**Pull request by nevi-me**
This is committed on top of the PR for the stream reader, and should be merged after it.

### CODE REVIEW

File: rust/arrow/src/ipc/writer.rs
Comment (nevi-me) [inline rust/arrow/src/ipc/writer.rs]: I need help here, I couldn't find a way of reusing the functions in `crate::ipc::convert`

**Comment by andygrove** [comment]
Hi @nevi-me I am back at work tomorrow and will start reviewing this.

File: rust/arrow/src/ipc/writer.rs
Comment (maxburke) [reply rust/arrow/src/ipc/writer.rs]: I pulled this change down and played with it; I'm pretty excited to get the IPC writer for Rust :).  I've run into similar challenges before with flatbuffers, and sometimes there's no good way of working around it. 

Would it be worth converting schema_to_fb_offset to a macro maybe?

Comment (nevi-me) [reply rust/arrow/src/ipc/writer.rs]: I can try the macro approach if I can't find another workaround, I could do this when working on the `StreamWriter` given that it'll share most code with the `FileWriter`.

**Comment by maxburke** [comment]
Hi! Is there anything I could be able to do to help with this merge?

**Comment by nevi-me** [comment]
Hey @maxburke, the next release is a few days away, so I doubt this will make the cut. If you need to use the IPC writer for now, you can work from my branch in the short term. I can keep this PR updated with any other Rust changes in master.

Do you also need a `StreamWriter`? I am holding off from working on it until we review and merge dictionary support and this file writer, but I could work on it if you need it sooner.

File: rust/arrow/src/ipc/writer.rs
Comment (nevi-me) [inline rust/arrow/src/ipc/writer.rs]: ended up not using this, will remove it as it duplicates `block_offsets`

**Comment by maxburke** [comment]
The StreamWriter would be great to have but I've mostly worked around it for now with the FileWriter :) 

I'm actually currently working from your branch already.

Really appreciate your help and progress on this!

**Review by andygrove — COMMENTED** [review]
This is looking great @nevi-me! What help do you currently need in order to merge this?

**Comment by nevi-me** [comment]
@andygrove I'm just waiting for a review

**Review by andygrove — APPROVED** [review]
LGTM

**Comment by andygrove** [comment]
@nevi-me After merging to master, I'm seeing a build failure in master:

```
failures:

---- ipc::writer::tests::read_and_rewrite_generated_files stdout ----
thread 'ipc::writer::tests::read_and_rewrite_generated_files' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 2, kind: NotFound, message: "No such file or directory" }', src/libcore/result.rs:1187:5

```

**Comment by nevi-me** [comment]
It's also happening on other PRs, I haven't had a chance to look at it. I'll look into it tomorrow morning

**Pull request closed without merging**
